File: RCH-playground/api-testing-suite/backend/utils/geo.py
"""
Geographic utilities
Common functions for geographic calculations
"""
import math
from typing import Optional


def calculate_distance_km(
    lat1: float,
    lon1: float,
    lat2: float,
    lon2: float
) -> float:
    """
    Calculate distance between two points using Haversine formula
    
    Args:
        lat1: Latitude of first point (degrees)
        lon1: Longitude of first point (degrees)
        lat2: Latitude of second point (degrees)
        lon2: Longitude of second point (degrees)
    
    Returns:
        Distance in kilometers (rounded to 2 decimal places)
    
    Raises:
        ValueError: If coordinates are invalid (out of range)
    """
    # Validate coordinates
    if not (-90 <= lat1 <= 90) or not (-90 <= lat2 <= 90):
        raise ValueError(f"Invalid latitude: must be between -90 and 90")
    if not (-180 <= lon1 <= 180) or not (-180 <= lon2 <= 180):
        raise ValueError(f"Invalid longitude: must be between -180 and 180")
    
    # Earth radius in kilometers
    R = 6371.0
    
    # Convert degrees to radians
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    
    # Haversine formula
    a = (
        math.sin(dlat / 2) ** 2 +
        math.cos(math.radians(lat1)) *
        math.cos(math.radians(lat2)) *
        math.sin(dlon / 2) ** 2
    )
    
    c = 2 * math.asin(math.sqrt(a))
    distance = R * c
    
    return round(distance, 2)


def calculate_distance_miles(
    lat1: float,
    lon1: float,
    lat2: float,
    lon2: float
) -> float:
    """
    Calculate distance between two points in miles
    
    Args:
        lat1: Latitude of first point (degrees)
        lon1: Longitude of first point (degrees)
        lat2: Latitude of second point (degrees)
        lon2: Longitude of second point (degrees)
    
    Returns:
        Distance in miles (rounded to 2 decimal places)
    """
    distance_km = calculate_distance_km(lat1, lon1, lat2, lon2)
    return round(distance_km * 0.621371, 2)


def validate_coordinates(
    lat: Optional[float],
    lon: Optional[float]
) -> bool:
    """
    Validate that coordinates are valid
    
    Args:
        lat: Latitude (degrees)
        lon: Longitude (degrees)
    
    Returns:
        True if coordinates are valid, False otherwise
    """
    if lat is None or lon is None:
        return False
    
    try:
        return (-90 <= lat <= 90) and (-180 <= lon <= 180)
    except (TypeError, ValueError):
        return False

File: RCH-playground/nextjs-migration/lib/data-engine/utils/geo.ts
/**
 * Geographic Utilities
 * Ported from Python utils/geo.py
 */

/**
 * Calculate distance between two points using Haversine formula
 * 
 * @param lat1 Latitude of first point (degrees)
 * @param lon1 Longitude of first point (degrees)
 * @param lat2 Latitude of second point (degrees)
 * @param lon2 Longitude of second point (degrees)
 * @returns Distance in kilometers (rounded to 2 decimal places)
 * @throws Error if coordinates are invalid
 */
export function calculateDistanceKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  // Validate coordinates
  if (!(-90 <= lat1 && lat1 <= 90) || !(-90 <= lat2 && lat2 <= 90)) {
    throw new Error('Invalid latitude: must be between -90 and 90');
  }
  if (!(-180 <= lon1 && lon1 <= 180) || !(-180 <= lon2 && lon2 <= 180)) {
    throw new Error('Invalid longitude: must be between -180 and 180');
  }

  // Earth radius in kilometers
  const R = 6371.0;

  // Convert degrees to radians
  const dLat = toRadians(lat2 - lat1);
  const dLon = toRadians(lon2 - lon1);

  // Haversine formula
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRadians(lat1)) *
      Math.cos(toRadians(lat2)) *
      Math.sin(dLon / 2) ** 2;

  const c = 2 * Math.asin(Math.sqrt(a));
  const distance = R * c;

  return Math.round(distance * 100) / 100; // Round to 2 decimal places
}

/**
 * Calculate distance between two points in miles
 * 
 * @param lat1 Latitude of first point (degrees)
 * @param lon1 Longitude of first point (degrees)
 * @param lat2 Latitude of second point (degrees)
 * @param lon2 Longitude of second point (degrees)
 * @returns Distance in miles (rounded to 2 decimal places)
 */
export function calculateDistanceMiles(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const distanceKm = calculateDistanceKm(lat1, lon1, lat2, lon2);
  return Math.round(distanceKm * 0.621371 * 100) / 100;
}

/**
 * Validate that coordinates are valid
 * 
 * @param lat Latitude (degrees)
 * @param lon Longitude (degrees)
 * @returns True if coordinates are valid, False otherwise
 */
export function validateCoordinates(
  lat: number | null | undefined,
  lon: number | null | undefined
): boolean {
  if (lat === null || lat === undefined || lon === null || lon === undefined) {
    return false;
  }

  try {
    return -90 <= lat && lat <= 90 && -180 <= lon && lon <= 180;
  } catch {
    return false;
  }
}

/**
 * Convert degrees to radians
 */
function toRadians(degrees: number): number {
  return degrees * (Math.PI / 180);
}



